Add HTTP-related utilties to the library

Avaq opened the pull request:
- [x] Implement HTTP related utilities
- [x] Add tests for said utilities
- [x] Update package description and tags
- [ ] Update repository description and tags

Avaq committed:
Add HTTP-related utilties to the library

Co-Authored-By: David Chambers <dc@davidchambers.me>

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+//. 2. log an error to the console if:
+//.     * A network error occurs.
+//.     * The reponse code is not 200.
+//.     * The JSON is malformed.
```
This capitalization and punctuation seems wrong to me. I suggest:

```javascript
//. 1. log `"Aldwin Vlasblom"` to the terminal if nothing weird happens; or
//. 2. log an error to the console if:
//.     * a network error occurs;
//.     * the response code is not 200; or
//.     * the JSON is malformed.
```

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+//. 1. Log `"Aldwin Vlasblom"` to the terminal if nothing weird happens; or
+//. 2. log an error to the console if:
+//.     * A network error occurs.
+//.     * The reponse code is not 200.
```
```suggestion
//.     * The response code is not 200.
```

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+//.
+//. This is the "lowest level" function for making HTTP requests. It does not
+//. handle buffering, encoding, content negotiation, or anything really.
+//. For most use-cases, you can use one of the more specialized functions:
```
```suggestion
//. For most use cases, you can use one of the more specialized functions:
```

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+//. handle buffering, encoding, content negotiation, or anything really.
+//. For most use-cases, you can use one of the more specialized functions:
+//.
+//. * [`send`](#sent): Make a generic HTTP request.
```
```suggestion
//. * [`send`](#send): Make a generic HTTP request.
```

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+
+//# messageMeta :: IncomingMessage -> MessageMeta
+//.
+//. Extracts the common meta-data from an [IncomingMessage][].
```
```suggestion
//. Extracts the common metadata from an [IncomingMessage][].
```

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+  headers: message.headers,
+});
+
+//# bufferResponse :: Charset -> IncomingMessage -> Future Error {meta :: MessageMeta, body :: String}
```
```suggestion
//# bufferResponse :: Charset -> IncomingMessage -> Future Error { meta :: MessageMeta, body :: String }
```

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+//. A version of [`buffer`](#buffer) specialized in [IncomingMessage][]s.
+//.
+//. Given a charset and an IncomingMessage, returns a Future with the
+//. [MessageMeta](#messagemeta) and the buffered, encoded, message body.
```
Is this meant to link to `MessageMeta` or to `messageMeta`?

Avaq replied:
It links to the `messageMeta` function, where readers will be able to find what the `MessageMeta` type is.

davidchambers replied:
Why not use `#messageMeta` as the fragment identifier? :)

Avaq replied:
I don't think that's supported on GitHub, right? Unless I create the anchor tag myself.

davidchambers replied:
Will you use Transcribe to generate the readme? If so, you will have a second anchor with the correctly capitalized identifier. :)

Avaq replied:
Oh. Right! :)

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+const defaultContentType = 'text/plain; charset=' + defaultCharset;
+
+// charsetRegex :: RegExp
+const charsetRegex = /charset=([^;\s]+)/;
```
This could produce false positives, I believe. Consider `Content-Type: xcharset=foo; charset=utf8`.

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+//. Returns a Future with the [MessageMeta](#messagemeta) and the buffered,
+//. encoded, message body.
+export const autoBufferResponse = message => {
+  const contentType = message.headers['content-type'] || defaultContentType;
```
Are you intentionally treating `''` as `undefined` here?

Avaq replied:
It doesn't matter because the code paths converge on the same result in the following statements. :man_shrugging:

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+//.
+//. Given a URL and a StrMap of request headers, returns a Future which
+//. makes a GET requests to the given resource, buffers its response and
+//. encodes it accordingly, falling back to utf8 encoding as described in
```
```suggestion
//. encodes it accordingly, falling back to UTF-8 encoding as described in
```

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+//.
+//. A version of [`request`](#request) for sending arbitrary data to a server.
+//.
+//. Given a mime type, a request method, a URL, a StrMap of headers, and
```
```suggestion
//. Given a MIME type, a request method, a URL, a StrMap of headers, and
```

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+//. Given a mime type, a request method, a URL, a StrMap of headers, and
+//. finally a Buffer, returns a Future which will send the Buffer to the
+//. server at the given URL using the given request method, telling it the
+//. buffer contains data of the given mime-type.
```
```suggestion
//. buffer contains data of the given MIME type.
```

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+//. [`responseToError`](#responsetoerror), you can then flatten it back into
+//. the outer Future.
+//.
+//. The usage example under the [Http](#http)-section shows this.
```
```suggestion
//. The usage example under the [Http](#http) section shows this.
```

davidchambers commented on index.js:
```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+//.   '  Dear user,\n' +
+//.   '  \n' +
+//.   '  Everything broke down.\n' +
+//.   '  We are sorry.\n'
```
Is the trailing `\n` intentional? The input lacks one.

Avaq replied:
Whoops.

Avaq commented:
Wow. Thank you for taking a look @davidchambers. I have incorporated your suggestions in ed28659

Avaq committed:
Update package description and add tags

The purpose of this is to improve findability now that
this can be used as a request library.

Avaq commented on index.js:
````diff
@@ -38,9 +58,50 @@ export const once = event => emitter => Future ((rej, res) => {
+//. 'Hello world!'
+//. ```
+export const encode = encoding => buffer => (
+  mapRej (e => new Error (e.message))
````
This is because `buffer.toString` throws some really weird doubly-wrapped NodeError of TypeError with extra properties, and I couldn't write an assertion against it. So I figured I'd save my users the trouble of dealing with that altogether. :)

Avaq commented:
I'm working on some small changes to the API which should improve overall versatility.

dicearr commented on index.js:
```diff
@@ -109,6 +182,271 @@ export const immediate = x => Future ((rej, res) => {
+//. provide a smoother usage experience.
+
+// getRequestModule :: String -> Future Error Module
+const getRequestModule = protocol => {
```
I'd rather align the type signature with the declaration when possible.

```js
//    getRequestModule :: String -> Future Error Module
const getRequestModule = protocol => {
```

dicearr commented on index.js:
````diff
@@ -109,6 +182,271 @@ export const immediate = x => Future ((rej, res) => {
+//.
+//. ```js
+//. retrieve ('https://api.github.com/users/Avaq')
+//.          ({headers: {'User-Agent': 'Avaq'}})
````
If I am not mistaken, the second argument should be an Object containing only the headers.

```js
retrieve ('https://api.github.com/users/Avaq')
         ({'User-Agent': 'Avaq'})
```

dicearr reviewed (commented):
Really looking forward to start using this! :rocket:

Avaq commented:
@dicearr [Feedback incorporated](https://github.com/fluture-js/fluture-node/compare/90e580920d061b02120fb3faa99f412cb4b3cf35..9163b899da675c088b67707594e22ffc52e63334), and rebased. I squashed the last commit with the first, as it was pretty much a rewrite of the first.
